### PR TITLE
Update PGA doc for UAST dataset

### PR DESCRIPTION
**Pull request by r0mainK**
@vmarkovtsev this is WIP, I left the ClickHouse part empty for now and only was as conservative as I could on the old doc. What do you think ?

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@r0mainK I have pushed my part directly to your fork and left a few TODOs in both files. PTAL

File: PublicGitArchive/pga/README.md [hunk 1]
```diff
@@ -16,20 +16,31 @@ In the meanwhile you'll need to compile this tool.
 
 There are three subcommands in `pga`: `list`, `get`, and `siva`.
 
+### Datasets
+
+Two datasets are exposed through this tool, and both the `list` and `get` command can be used to explore and retrieve them. To do so, you must specify with a keyword which dataset you want to work on :
+
+- `siva`: The original Public Git Archive dataset, made up of Siva files.
```
Comment (vmarkovtsev) [inline PublicGitArchive/pga/README.md]:
```suggestion
- `siva`: The original Public Git Archive dataset, made of Siva files.
```
to make up = to imagine and tell stories :D

File: PublicGitArchive/pga/README.md [hunk 2]
```diff
@@ -16,20 +16,31 @@ In the meanwhile you'll need to compile this tool.
...
+- `uast`: The [dataset](../../PublicGitArchiveUASTs) created by extracting UASTs from the HEAD commit of each repository, made up of Parquet files.
+
+Note that the `siva` _command_ does not work on Parquet files.
```
Comment (vmarkovtsev) [inline PublicGitArchive/pga/README.md]:
```suggestion
Note that the `siva` _command_ does not work with Parquet files.
```

File: PublicGitArchive/pga/README.md [hunk 3]
```diff
@@ -16,20 +16,31 @@ In the meanwhile you'll need to compile this tool.
...
+
 ### Listing repositories
 
 When you run `pga list` two things wil happen.
-First a copy of the latest index for the Public Git Archive will be downloaded and cached locally.
+First a copy of the latest index for the dataset specified will be downloaded and cached locally.
```
Comment (vmarkovtsev) [inline PublicGitArchive/pga/README.md]:
```suggestion
First a copy of the latest index for the specified dataset will be downloaded and cached locally.
```

File: PublicGitArchiveUASTs/ClickHouse/README.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+The procedure took 3 days on 16 2-vcore, 16GB worker instances and a 64-vcore, 58GB server instance on
+Google Cloud.
+
+### Limitations
+
+As was already mentioned in the [schema section](#schema), the UASTs are flattened in a lossy way.
+This includes the aggressive normalization of the data for each programming language. While we did our
+best at detecting the possible problems early on, a few sneaked inside when it was too late to fix them. Furthermore, the dataset was not updated after the 21st of september, so any commit merged after will not have been included.
+
+Below is a non-exhaustive list of known issues:
```
Comment (vmarkovtsev) [inline PublicGitArchiveUASTs/ClickHouse/README.md]:
```suggestion
Below is an incomplete list of known issues:
```

File: PublicGitArchiveUASTs/ClickHouse/README.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+* Duplication: 
+    * Some repositories were processed multiple times, because they appeared in multiple Parquet files. We created the DB by iterating on each Parquet file, and did not check for duplicate rows.
+    * Errors inherited from Babelfish (language specific, e.g. [duplicate go comments](https://github.com/bblfsh/go-driver/issues/56)).
```
Comment (vmarkovtsev) [inline PublicGitArchiveUASTs/ClickHouse/README.md]:
```suggestion
    * Errors inherited from Babelfish that are language-specific, e.g. [duplicate go comments](https://github.com/bblfsh/go-driver/issues/56).
```

File: PublicGitArchiveUASTs/ClickHouse/README.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+* Missing data:
+    * Due to how we traversed the trees, in some cases data was purged, e.g. all Ruby imports. See [here](https://github.com/src-d/uast2clickhouse/issues/11).
+    * Errors inherited from Babelfish (language specific, e.g. some drivers do not keep keywords).
+* Wrong ordering: the left and right columns tend to be unreliable, due to how we traversed the UASTs and propagated positional information  - the line number is correct though.
```
Comment (vmarkovtsev) [inline PublicGitArchiveUASTs/ClickHouse/README.md]:
```suggestion
* Wrong ordering: the left and the right columns are sometimes unreliable, due to how we traversed the UASTs and propagated positional information. The line numbers should always be correct though.
```

File: PublicGitArchiveUASTs/README.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+- `head` (string): the UUID of the repository of the given file. You can use [this mapping](https://drive.google.com/open?id=136vsGWfIwfd0IrAdfphIU6lkMmme4-Pj)
+to obtain the repository names from UUIDs.
+- `path` (string): the filepath to the given file, in the repository structure.
+- `uast` (variable-length byte array): the UAST of the given file.
+
+### Usage
+
+The Parquet files can be read using any library that supports the format, however using Spark is strongly advised if processing a large part of the dataset. The UASTs are stored as byte arrays, and thus you can use any of the [Babelfish Clients](https://doc.bblf.sh/using-babelfish/clients.html) to read and manipulate them.
+
+As an example, here is how one would extract all identifiers from the UASTs in a given Parquet file:
```
Comment (vmarkovtsev) [inline PublicGitArchiveUASTs/README.md]:
```suggestion
For example, this is how to extract all identifiers from the UASTs in a given Parquet file:
```

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: PublicGitArchive/README.md [hunk 8]
```diff
@@ -13,6 +13,8 @@ Besides, there is a number of auxiliary datasets:
 * [configs.tar.xz](https://drive.google.com/open?id=1_cij4BMrPiKVBVdZyUzg1iOhB3pL6EPR) - raw git config files for each siva.
 * [heads.csv.xz](https://drive.google.com/open?id=136vsGWfIwfd0IrAdfphIU6lkMmme4-Pj) - mapping from HEAD UUID to repository name.
 
+Since the second version of PGA, we have also created a [dataset of UASTs](../PublicGitArchiveUASTs), obtained from files in the HEAD commit of each repository.
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchive/README.md]: done

File: PublicGitArchive/pga/README.md [hunk 9]
```diff
@@ -16,20 +16,31 @@ In the meanwhile you'll need to compile this tool.
 
 There are three subcommands in `pga`: `list`, `get`, and `siva`.
 
+### Datasets
+
+Two datasets are exposed through this tool, and both the `list` and `get` command can be used to explore and retrieve them. To do so, you must specify with a keyword which dataset you want to work on :
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchive/pga/README.md]: done

File: PublicGitArchive/pga/README.md (hunk 1 above)
Comment (r0mainK) [reply PublicGitArchive/pga/README.md]: done :p

File: PublicGitArchive/pga/README.md [hunk 10]
```diff
@@ -16,20 +16,31 @@ In the meanwhile you'll need to compile this tool.
...
+
+- `siva`: The original Public Git Archive dataset, made up of Siva files.
+- `uast`: The [dataset](../../PublicGitArchiveUASTs) created by extracting UASTs from the HEAD commit of each repository, made up of Parquet files.
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchive/pga/README.md]: done

File: PublicGitArchive/pga/README.md (hunk 2 above)
Comment (r0mainK) [reply PublicGitArchive/pga/README.md]: done

File: PublicGitArchive/pga/README.md (hunk 3 above)
Comment (r0mainK) [reply PublicGitArchive/pga/README.md]: done

File: PublicGitArchiveUASTs/ClickHouse/README.md (hunk 4 above)
Comment (r0mainK) [reply PublicGitArchiveUASTs/ClickHouse/README.md]: done

File: PublicGitArchiveUASTs/ClickHouse/README.md (hunk 5 above)
Comment (r0mainK) [reply PublicGitArchiveUASTs/ClickHouse/README.md]: done

File: PublicGitArchiveUASTs/ClickHouse/README.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+This includes the aggressive normalization of the data for each programming language. While we did our
+best at detecting the possible problems early on, a few sneaked inside when it was too late to fix them. Furthermore, the dataset was not updated after the 21st of september, so any commit merged after will not have been included.
+
+Below is a non-exhaustive list of known issues:
+
+* Duplication: 
+    * Some repositories were processed multiple times, because they appeared in multiple Parquet files. We created the DB by iterating on each Parquet file, and did not check for duplicate rows.
+    * Errors inherited from Babelfish (language specific, e.g. [duplicate go comments](https://github.com/bblfsh/go-driver/issues/56)).
+* Missing data:
+    * Due to how we traversed the trees, in some cases data was purged, e.g. all Ruby imports. See [here](https://github.com/src-d/uast2clickhouse/issues/11).
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchiveUASTs/ClickHouse/README.md]: done

File: PublicGitArchiveUASTs/ClickHouse/README.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    * Errors inherited from Babelfish (language specific, e.g. some drivers do not keep keywords).
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchiveUASTs/ClickHouse/README.md]: done

File: PublicGitArchiveUASTs/ClickHouse/README.md (hunk 6 above)
Comment (r0mainK) [reply PublicGitArchiveUASTs/ClickHouse/README.md]: done

File: PublicGitArchiveUASTs/ClickHouse/README.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+* Wrong ordering: the left and right columns tend to be unreliable, due to how we traversed the UASTs and propagated positional information  - the line number is correct though.
+
+If you see something strange or have trouble using the dataset, please say so [here](https://github.com/src-d/uast2clickhouse/issues) and we will look into it - however the dump will not be updated until Public Git Archive v3 is released.
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchiveUASTs/ClickHouse/README.md]: done

File: PublicGitArchiveUASTs/README.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+=======================================
+
+The [Universal Abstract Syntax Trees](https://doc.bblf.sh/uast/uast-specification-v2.html) (UASTs) extracted
+from the latest (HEAD) revision of every Git reference contained in [Public Git Archive](../../PublicGitArchive).
+The dataset is distributed as Parquet files, which you can download using the [pga CLI](../PublicGitArchive/pga).
+There is also a [ClickHouse DB version](ClickHouse) which is more lightweight and easier to work with.
+
+### Format
+
+The Parquet files have 3 columns (each row is a single file):
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchiveUASTs/README.md]: done

File: PublicGitArchiveUASTs/README.md [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+- `head` (string): the UUID of the repository of the given file. You can use [this mapping](https://drive.google.com/open?id=136vsGWfIwfd0IrAdfphIU6lkMmme4-Pj)
+to obtain the repository names from UUIDs.
+- `path` (string): the filepath to the given file, in the repository structure.
+- `uast` (variable-length byte array): the UAST of the given file.
+
+### Usage
+
+The Parquet files can be read using any library that supports the format, however using Spark is strongly advised if processing a large part of the dataset. The UASTs are stored as byte arrays, and thus you can use any of the [Babelfish Clients](https://doc.bblf.sh/using-babelfish/clients.html) to read and manipulate them.
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchiveUASTs/README.md]: done

File: PublicGitArchiveUASTs/README.md (hunk 7 above)
Comment (r0mainK) [reply PublicGitArchiveUASTs/README.md]: done

File: PublicGitArchiveUASTs/README.md [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+As an example, here is how one would extract all identifiers from the UASTs in a given Parquet file:
+
+```
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchiveUASTs/README.md]: done

File: PublicGitArchiveUASTs/README.md [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+We used 11 "Start-2-L" machines on online.net.
+
+### Limitations
+
+|           | # of repos | # of files | # of distinct files | % of duplicates |
+|:---------:|:----------:|:----------:|:-------------------:|:---------------:|
+| **PGA**   | 220,174    | 40,971,787 | 40,829,244          | 0.3 %           |
+| **UASTs** | 218,023    | 36,162,330 | 35,991,340          | 0.5 %           |
+
+As the table above shows, we were not able to process 100% of the HEAD of Public Git Archive. For one, we could not process all languages, as Babelfish currently only has drivers for 9 languages. Furthermore, some files proved too large to be processable in a reasonnable amount of time. This, combined to parsing errors or bugs on Babelfish's side, resulted in missing ~12% of all parsable files in the HEAD of PGA, amounting for ~45% of the data in bytesizes. As we can see from the table below, not all languages all equal: for instance, the C++ driver, which handles all C-like languages (C, C++, Metal, Cuda), performed much worse then most lamguages, while the Go driver performed much better then the rest of the languages.
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchiveUASTs/README.md]: done

File: PublicGitArchiveUASTs/README.md [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+# We apply the pipeline, then trigger execution with `show`
+
+df = spark.read.parquet("/path/to/parquet")
+df = df.withColumn("identifier", explode(extract_identifiers_udf(df.uast))) \
+    .select("head", "path", "identifier")
+df.show()
+```
+
+Please note that the [Babelfish Python Client](https://github.com/bblfsh/python-client) needs to be present on the Spark workers for this snippet to work, **not only on the driver.**
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply PublicGitArchiveUASTs/README.md]: done

**Review by vmarkovtsev — APPROVED** [review]

**Pull request merged**